core/src/main/scala/kafka/server/ReplicationQuotaManager.scala

```diff
@@ -0,0 +1,219 @@
+  *
+  * http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+package kafka.server
+
+import java.util.concurrent.{ConcurrentHashMap, TimeUnit}
+import kafka.common.TopicAndPartition
+import kafka.server.Constants._
+import kafka.server.ReplicationQuotaManagerConfig._
+import kafka.utils.CoreUtils._
+import kafka.utils.{CoreUtils, Logging}
+import org.apache.kafka.common.MetricName
+import org.apache.kafka.common.metrics._
+import java.util.concurrent.locks.ReentrantReadWriteLock
+import org.apache.kafka.common.metrics.stats.{Rate, SimpleRate}
+import org.apache.kafka.common.utils.Time
+
+/**
+  * Configuration settings for quota management
+  *
+  * @param quotaBytesPerSecondDefault The default bytes per second quota allocated to internal replication
+  * @param numQuotaSamples            The number of samples to retain in memory
+  * @param quotaWindowSizeSeconds     The time span of each sample
+  *
+  */
+case class ReplicationQuotaManagerConfig(quotaBytesPerSecondDefault: Long = QuotaBytesPerSecondDefault,
+                                         numQuotaSamples: Int = DefaultNumQuotaSamples,
+                                         quotaWindowSizeSeconds: Int = DefaultQuotaWindowSizeSeconds)
+
+object ReplicationQuotaManagerConfig {
+  val QuotaBytesPerSecondDefault = Long.MaxValue
+  // Always have 10 whole windows + 1 current window
+  val DefaultNumQuotaSamples = 11
+  val DefaultQuotaWindowSizeSeconds = 1
+  // Purge sensors after 1 hour of inactivity
+  val InactiveSensorExpirationTimeSeconds = 3600
+}
+
+trait ReadOnlyQuota {
+  def bound(): Int
+
+  def isThrottled(topicAndPartition: TopicAndPartition): Boolean
+
+  def isQuotaExceededBy(bytes: Int): Boolean
+}
+
+object Constants {
+  val allReplicas = Seq[Int](-1)
+}
+
+/**
+  * Tracks replication metrics and comparing them to any quotas for throttled partitions.
+  *
+  * @param config          The quota configs
+  * @param metrics         The Metrics instance
+  * @param replicationType The name / key for this quota manager, typically leader or follower
+  * @param time            Time object to use
+  */
+class ReplicationQuotaManager(val config: ReplicationQuotaManagerConfig,
+                              private val metrics: Metrics,
+                              private val replicationType: String,
+                              private val time: Time) extends Logging with ReadOnlyQuota {
+
+  private val lock = new ReentrantReadWriteLock()
+  private val throttledPartitions = new ConcurrentHashMap[String, Seq[Int]]()
+  private var quota: Quota = null
+  private val sensor = createSensor(quota, replicationType)
```

Comment:
This may not be enough since the sensor can be removed after it's inactive for some time (say, throttled replicas are removed). When that happens, sensor will be pointing to an obsolete object. To be safe, we probably need a getOrCreateQuotaSensors() method like in ClientQuotaManager.